misc/mdns/src/service.rs

```diff
@@ -173,39 +190,70 @@ impl MdnsService {
         // Flush the send buffer of the main socket.
         while !self.send_buffers.is_empty() {
             let to_send = self.send_buffers.remove(0);
-            match self.socket.send_to(&to_send, &From::from(([224, 0, 0, 251], 5353))).await {
-                Ok(bytes_written) => {
+
+            // TODO: The future borrows `to_send`. We can't insert `to_send`
+            // back into `send_buffers` before the borrow is dropped. We work
+            // around this with the additional scope on match. Is there a
+            // prettier way?
+            match {
+                let fut = self.socket.send_to(&to_send, *IPV4_MDNS_MULTICAST_ADDRESS);
+                // TODO: Is it safe to pin the future on the stack?
+                futures::pin_mut!(fut);
+                // TODO: Is it safe to drop the future on pending and create a new one later?
+                fut.poll(cx)
+            } {
+                Poll::Ready(Ok(bytes_written)) => {
                     debug_assert_eq!(bytes_written, to_send.len());
                 }
-                Err(_) => {
+                Poll::Ready(Err(_)) => {
                     // Errors are non-fatal because they can happen for example if we lose
                     // connection to the network.
                     self.send_buffers.clear();
                     break;
                 }
+                Poll::Pending => {
+                    self.send_buffers.insert(0, to_send);
+                    break;
+                }
             }
         }
 
         // Flush the query send buffer.
         // This has to be after the push to `query_send_buffers`.
         while !self.query_send_buffers.is_empty() {
             let to_send = self.query_send_buffers.remove(0);
-            match self.socket.send_to(&to_send, &From::from(([224, 0, 0, 251], 5353))).await {
-                Ok(bytes_written) => {
+
+            // TODO: The future borrows `to_send`. We can't insert `to_send`
+            // back into `send_buffers` before the borrow is dropped. We work
+            // around this with the additional scope on match. Is there a
+            // prettier way?
+            match {
+                let fut = self.query_socket.send_to(&to_send, *IPV4_MDNS_MULTICAST_ADDRESS);
```

Comment:
That doesn't work, you're dropping the future. There's a lot more code to write here for it to work unfortunately.